packages/blade/src/utils/useBreakpoint/useBreakpoint.ts

```diff
@@ -0,0 +1,144 @@
+import { useEffect, useState, useCallback, useMemo } from 'react';
+import type { Breakpoints } from '../../tokens/global';
+
+type DeviceType = 'desktop' | 'mobile';
+type Breakpoint = keyof Breakpoints | undefined;
+
+type BreakpointAndDevice = {
+  matchedBreakpoint: Breakpoint;
+  matchedDeviceType: DeviceType;
+};
+
+const useBreakpoint = ({ breakpoints }: { breakpoints: Breakpoints }): BreakpointAndDevice => {
+  const supportsMatchMedia =
+    typeof document !== 'undefined' && typeof window?.matchMedia === 'function';
+
+  const breakpointsTokenAndQueryCollection = useMemo(
+    () =>
+      (supportsMatchMedia
+        ? Object.entries(breakpoints).map(
+            ([breakpointTokenName, breakpointSize], index, breakpointsArray) => {
+              let mediaQuery = '';
+
+              if (breakpointTokenName === 'max') {
+                mediaQuery = `screen and (min-width: ${breakpointSize}px)`;
+              } else if (breakpointsArray[index - 1]) {
+                mediaQuery = `screen and (min-width: ${
+                  breakpointsArray[index - 1][1] + 1
+                }px) and (max-width: ${breakpointSize}px)`;
+              } else {
+                mediaQuery = `screen and (max-width: ${breakpointSize}px)`;
+              }
+
+              return [breakpointTokenName, breakpointSize, mediaQuery];
+            },
+          )
+        : []) as [keyof Breakpoints, number, string][],
+    [breakpoints, supportsMatchMedia],
+  );
+
+  const getMatchedDeviceType = useCallback((matchedBreakpoint: Breakpoint): DeviceType => {
+    let matchedDeviceType: DeviceType = 'mobile';
+    if (typeof navigator !== 'undefined' && navigator.product === 'ReactNative') {
+      // react-native
+      matchedDeviceType = 'mobile';
+    } else if (typeof document !== 'undefined') {
+      // browser
+      if (matchedBreakpoint && ['xs', 's', 'm'].includes(matchedBreakpoint)) {
+        // tablet is also categorised as mobile
+        matchedDeviceType = 'mobile';
+      } else {
+        matchedDeviceType = 'desktop';
+      }
+    } else if (typeof process !== 'undefined') {
+      // node
+      //@TODO: Check for useragent for node
+      matchedDeviceType = 'desktop';
+    }
+    return matchedDeviceType;
+  }, []);
+
+  const getMatchedBreakpoint = useCallback(
+    (event?: MediaQueryListEvent): Breakpoint => {
+      const matchedBreakpoint =
+        breakpointsTokenAndQueryCollection.find(([_, __, mediaQuery]) => {
+          if (event?.media === mediaQuery) {
+            return true;
+          }
```

Comment:
this will fail when we the state is initialised for the first time when the hook is getting initialised and hence the event object will be empty so we'll fallback to browser's window object